database.hh

```diff
@@ -1013,50 +1013,6 @@ public:
     friend class distributed_loader;
 };
 
-using sstable_reader_factory_type = std::function<flat_mutation_reader(sstables::shared_sstable&, const dht::partition_range& pr)>;
```

Comment:
The idea is that different sstable sets have different ways, specific to the sets' implementations, to efficiently query the set. Only the sstable set knows how to do that. Whether it's by sophisticated merging of the sstables with incremental opening or whatever, it's an implementation detail of the set.

Sometimes there is no suitable abstract interface, because the problems being solved in the different cases are different.
Of course I could provide an interface e.g. that abstracts over the `min_position_reader_queue`, for incrementally opening sstables based on position_in_partition. But that interface would be useless because it is created with one specific sstable set in mind, and the others you must artificially force to implement that interface, and of course the other implementations would be just generic "return all sstables in the partition". It's a [false abstraction](https://mortoray.com/2014/08/01/the-false-abstraction-antipattern/).

> This is making our already very bad everything-depends-on-everything even worse

`FragmentProducer` and `mutation_fragment_merger` are, I believe, useful general abstractions that can be reused in multiple places. So maybe it would be even better to move them to a separate header just for them. But definitely I think they should be accessible to the "general public".

The alternative is to have the `mutation_reader` module expose public interfaces to specific readers that are mapped one-to-one to some specific code in other places of the codebase where they are used to solve a single problem. **This** is making everything-depend-on-everything: we're making the logic of a single algorithm used to solve a specific problem scattered all over the place by creating false leaky interfaces. When one reads the code and sees the interfaces one may think that this is a general abstraction which you can plug and use in multiple places, but there is strong hidden coupling. And that's what's very bad.

Reply:
There are only so many different ways to query an sstable set. Query is the wrong word really because what we are doing is selecting the sstables we need to read for a given read-range. And there is not an infinite variety on how one can do that. Really, I cannot think of any other way then select-all or select incrementally based on the position. And there is a simple way to abstract these two ways -- the selector (or the `min_position_reader_queue` as you call it).

What all this machinery should be is a mirror of the existing combined reader, just operating on the clustering key level, instead of on the ring-position level. The two problems are *very* similar. Having two vastly different solutions to almost the same problem makes the mental load of understanding and maintaining this code much higher.

Reply:
> There are only so many different ways to query an sstable set. Query is the wrong word really because what we are doing is selecting the sstables we need to read for a given read-range. And there is not an infinite variety on how one can do that.

The problem I'm trying to solve here is not to select sstables from a set for the given range; the problem is to read mutations from a set of sstables efficiently. The range-selecting problem is a sub-problem of the read problem, introduced in order to improve performance. It's an implementation detail of the read problem *for a specific type of an sstable set*.

Given a concrete sstable set, there are only so many different ways to read it -- I agree. For example, suppose you want to read a range of partitions. For `partitioned_sstable_set`, you can read it by opening all readers at the beginning, or by incrementally opening them as you proceed through the partition range. `partitioned_sstable_set` makes the second option possible *because* it's "partitioned".

`bag_sstable_set` does not have that second option. You can pretend it does e.g. by introducing an interface with a `select(range)` method, but that doesn't change facts --- this method will just return all sstables every time you call it, and you know that because `bag_sstable_set` does not guarantee any structure on the set of sstables, it's just a "bag". The only way to read a `bag_sstable_set` is to open all readers at the beginning. The second option of incrementally selecting sstables is actually equal to the first option, you just gave two different names to the same thing. There is nothing "incremental" about how you read a `bag_sstable_set`.

The point I'm trying to make: *different types of sstable sets have different ways of reading them*. One set offers a way to incrementally open sstables as you proceed through partitions in a range. Another set offers a way to incrementally open sstables as you proceed through positions in a single partition. Yet another set does not provide any of this, it only understands how to open all sstables at once.

I can think of new types of sstable sets with different guarantees, and with specific ways (specific only to these types) to query them more efficiently by using the additional assumptions. For example, imagine an sstable set that guarantees that every sstable contains a single partition. That is, a given partition can appear in multiple sstables, but there cannot be two different partitions in a single sstable (I know it would be a bad idea to do that in general, but maybe for use cases where there is a very small number of partitions). Now when you write code for querying a partition range for this sstable set, you can use better algorithms and data structures that take advantage of the assumption to provide better efficiency compared to the other sstable sets.

I don't want to introduce an interface abstracting over `min_position_reader_queue`, because it's fake. We would pretend that other sstable set types provide an efficient way of incrementally opening sstables when reading a partition, but that's it -- it's just pretending. All sstables probably overlap in their clustering ranges and most of the time we would be opening all of them at once, and we're back to the `bag` option. It's a false abstraction.

> The two problems are very similar.

There are significant differences. Opening new readers in the middle of a partition, after you've already returned some fragments from the partition, is not the same as opening new readers at partition boundaries. It requires additional assumptions such as no partition tombstones (which are stored in the `partition-start` fragments) or static rows (which come as the first fragment after `partition-start`); those are very specific conditions. You also want to close the readers in middle of a partition, which becomes non-trivial to do efficiently in presence of fast-forwarding. So the problems may be similar at first glance, but the devil lies in the details; it turns out that solving the two problems require different methods.

Reply:
I disagree. Reading from all sstable sets is the same. You read from N sstables at a time, you merge their output and that's it. There is no sstable-set specific thing here. In fact, up to a few years ago we had a single algorithm for reading from all the different sstable sets and it was fine and it worked.

The only thing different between two sstable sets is the layout of the sstables. This can allow for certain optimizations, namely to add certain sstables to the read later and/or to remove them earlier, thus reducing the amount of sstables read in parallel. There is a simple abstraction for this: the read represents an iterator or cursor, that traverses the read-range fragment-by-fragment. We have to open all sstables that have data intersecting with this cursor.

Now certain sstable sets have specific layouts with properties that allow us to know in advance that not all sstables will intersect with the cursor at all times. There are these two options: either all sstables intersect at all times or the fact that they don't is just by chance; or they are partitioned in some way. So we have two discreet options: non-incremental and incremental. Furthermore, incremental can be incremental on the clustering row level, or incremental on the partition level. This can be abstracted by two selectors: one that selects on the partition level, and one that selects on the clustering level.

The point I'm trying to make is that the problem space is not infinite, there is not an arbitrary amount of ways to read from sstable sets. There is just one single way to read them, and 3 different planes of incremental selection they support:
* none (STCS)
* partition-level (LCS, ICS)
* clustering-level (TWCS)

As our data-model has just two planes of sorting data, there is not an arbitrary other possible ways to do the selection. The single-partition sstable is just a sub-case of the partition-level incremental selection.

Just for posterity, there is yet another plane of sstable exclusion: timestamp. At one point we were working on a special reader for single row reads that would exclude sstables if it could prove that they have only older data than what the reader already had. But this didn't use any sstable set specific things.

Reply:
Agree with @denesb. We created incremental_selector as a way for sstable_set to interact with sstable_readers.

However, to keep an open mind, if it turns out that we need a new incremental_selector *interface* for different selection optimizations, then maybe it's not such a good interface. Is this the case?

Reply:
Creating a new interface is exactly what I'm trying to avoid here.
The current state of things is as follows:
1. `time_series_reader_merger` uses `min_position_reader_queue`
2. `min_position_reader_queue` knows `time_series_sstable_set` internals (specifically, the type of the container that stores the sstables)
3. therefore, `time_series_reader_merger` is specific to `time_series_sstable_set`

We could introduce an `abstract_min_position_reader_queue` interface (corresponding to `incremental_selector`, but for clustering keys) and have `time_series_reader_merger` use this abstract interface. 

But my problem with this is that `abstract_min_position_reader_queue` is a "false abstraction". All sstable sets would have to implement it "just because". The only "real" implementation would be `min_position_reader_queue`, other sstable sets would have a fake implementation that returns all sstables.

Reply:
@denesb 
> Reading from all sstable sets is the same. 

The argument you give proves that it's not. They have different methods of "incremental selection":

> 3 different planes of incremental selection they support

that's just saying "different algorithms for reading them" with extra steps.

---

My suggestion is as follows: do not create another "incremental selection" interface **for now**.

If a new sstable set type appears in the future, which also supports efficient clustering-based incremental selection, the code can be **very easily** refactored by introducing such an interface, which `min_position_reader_queue` will be a subclass of and `time_series_reader_merger` will use. It's a really quick refactor. And we will get a clearer picture of how the interface should look like, because it will now have 2 real implementations.

On the other hand, refactoring is much harder when we introduce a bunch of interfaces prematurely.

Reply:
Interfaces are also used to provide a clean separatation of code that logically belong to different modules. In this quality they might have a single implementation, not even one real + many "false" ones. The existing incremental selector is one such interface -- it is not really abstracting anything in the polymorphism sense, it just provides a clean separation between iterating the sstable set and how the readers are merged, without the two parties having to know about each other.
This patch-set has to solve the same problem and I expected the same solution. Not because it is the only solution or even necessarily the very best one, but because it is something we are already familiar with and its nice to be consistent (not at all cost of course).

Reply:
Maybe eliminating the intermediary has merit. If every type of incremental selection strategy requires a new intermediary (because the incremental selection is keyed off a different item) it indicates that the interface is not very good.

Reply:
But, what about memtables? Here we have something that doesn't belong in an sstable_set, yet can participate in selection.